### PR TITLE
travis: enable Travis CI on CentOS 7

### PR DESCRIPTION
**Pull request by mrc0mmand**
Let's enable Travis CI on CentOS 7 for the systemd RHEL fork.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
So, the Travis itself is working fine, yet the test suite doesn't like the docker environment.

```
FAIL: test-path
===============
Cannot add dependency job for unit local-fs.target, ignoring: Unit not found.
Cannot add dependency job for unit swap.target, ignoring: Unit not found.
Cannot add dependency job for unit slices.target, ignoring: Unit not found.
Cannot add dependency job for unit swap.target, ignoring: Unit not found.
Cannot add dependency job for unit local-fs.target, ignoring: Unit not found.
Cannot add dependency job for unit slices.target, ignoring: Unit not found.
Test timeout when testing path-changed.path
path-service.service: state = dead; result = success 
path-service.service: state = start; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = start; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
path-service.service: state = dead; result = success 
FAIL: test-path-util
====================
---/a/b/c/d---
---/a/b/c---
---/a/b---
---/a---
------
---/a/b/c---
---/a/b---
---/a---
------
Assertion 'rlf == 0' failed at src/test/test-path-util.c:155, function test_path_is_mount_point(). Aborting.
expected: ./aa/bb/..
actual: ./aa/bb/..
expected: /aa///
actual: /aa
expected: /aa///
actual: /aa
expected: error
actual: error
/bin/sh
/build/./test-path-util
/usr/bin/sh
/bin/sh
/build/./test-path-util
/usr/bin/sh
FAIL: test-capability
=====================
Capabilities:= cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,35,36,37+eip
Capabilities:= cap_dac_override,cap_net_raw+ep
Capabilities:= cap_dac_override+ep
Failed to drop auxiliary groups list: Operation not permitted
Failed to change group ID: Operation not permitted
Assertion 'fv == CAP_CLEAR' failed at src/test/test-capability.c:164, function test_update_inherited_set(). Aborting.
FAIL: test-copy
===============
Assertion 'access(f, F_OK) == 0' failed at src/test/test-copy.c:114, function test_copy_tree(). Aborting.
```

We dealt with the `test-capability` fail already in upstream (https://github.com/systemd/systemd/issues/10663), fix is https://github.com/systemd/systemd/commit/c446b8486d9ed18d1bc780948ae9ee8a53fa4c3f. I'll look for fixes for the rest of the fails as well. 

Also, the cherry-picked commits are just for testing purposes, will drop them later. I'm not sure what workflow should be used for merging them, cc @lnykryn, @msekletar.

**Comment by mrc0mmand** [comment]
Fix for `test-path-util` is https://github.com/systemd/systemd/commit/be24321f3dae91a166166b239954032727439942. As the source tree is significantly changed in this version, it needs to be picked manually (in our tree it's `src/shared/path-util.c`). Also, it's not a test-only change, so it should get a more thorough review.

Issue: https://github.com/systemd/systemd/issues/11092
PR: https://github.com/systemd/systemd/pull/11094

**Comment by mrc0mmand** [comment]
Fix for `test-copy` is systemd/systemd@ef202b848bb6635dec17d3ec0041b04cd2301bed - again, this patch must be rewritten for our source tree. This patch changes the original behavior of the `fd_copy_directory` function (see the commit description), so we should take this into consideration. Also, the related PR contains another two commits, which should be backported as well.

PR: https://github.com/systemd/systemd/pull/9213/commits

**Comment by mrc0mmand** [comment]
`test-path` will apparently need some debugging, as I can't find any test-related change in upstream, which would fix the timeout and I can't even reproduce it *manually*:

```
# dnf -y install docker
# systemctl start docker
# git clone https://github.com/lnykryn/systemd-rhel
# git fetch -fu origin refs/pull/259/merge:pr
# git checkout pr
# export CONT_NAME="systemd-debug"
# ci/travis-centos.sh SETUP RUN
...
PASS: test-path
...
============================================================================
Testsuite summary for systemd 219
============================================================================
# TOTAL: 106
# PASS:  101
# SKIP:  5
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

File: ci/travis-centos.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    case $phase in
+        SETUP)
+            info "Setup phase"
+            info "Using Travis $CENTOS_RELEASE"
+            # Pull a Docker image and start a new container
+            docker pull centos:$CENTOS_RELEASE
+            info "Starting container $CONT_NAME"
+            $DOCKER_RUN -v $REPO_ROOT:/build:rw \
+                        -w /build --privileged=true --name $CONT_NAME \
+                        -dit --net=host centos:$CENTOS_RELEASE /sbin/init
```
Comment (msekletar) [inline ci/travis-centos.sh]: I am no expert on docker but I think that ```-d``` and ```-i``` are contradictory. I think you should drop ```it``` part.

File: ci/travis-centos.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            # Beautiful workaround for Fedora's version of Docker
```
Comment (msekletar) [inline ci/travis-centos.sh]: I am not quite sure I get this Fedora reference. After all Travis CI is running on Ubuntu, is it not?

File: ci/travis-centos.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            sleep 1
+            $DOCKER_EXEC yum makecache
+            # Install necessary build/test requirements
+            $DOCKER_EXEC yum -y --exclude selinux-policy\* upgrade
```
Comment (msekletar) [inline ci/travis-centos.sh]: This is weird as both ```centos:latest``` and ```centos:7``` don't even contain selinux-policy package.

**Review by msekletar — COMMENTED** [review]
Otherwise looks good, thank you!

File: ci/travis-centos.sh (hunk 2 above)
Comment (mrc0mmand) [reply ci/travis-centos.sh]: The `travis-centos.sh` script is a slightly modified version of the Fedora "manager" we implemented in upstream (https://github.com/systemd/systemd/blob/master/travis-ci/managers/fedora.sh). One of the goals was to ensure local reproducibility, i.e. you can simply execute:
```
# dnf -y install docker
# systemctl start docker
# export CONT_NAME="my-debug-container"
# ci/travis-centos.sh SETUP RUN
```

and you can easily debug the container locally. Unfortunately, Docker shipped with Fedora contains a race condition, so we just put sleep here to sort of workaround it (the issue hasn't been solved yet).

File: ci/travis-centos.sh (hunk 3 above)
Comment (mrc0mmand) [reply ci/travis-centos.sh]: Good point, that's a leftover from the Fedora version of this script, will fix.

File: ci/travis-centos.sh (hunk 1 above)
Comment (mrc0mmand) [reply ci/travis-centos.sh]: Even though logically they are sort of contradictory, in practice you need to force-create a pseudo-terminal when the container is started even if you detach it immediately. I vaguely remember this had an actual use case in the upstream, but can't find it. However, there are plenty of other *explanations", e.g. https://stackoverflow.com/questions/41916435/practically-what-is-the-difference-between-docker-run-dit-itd-vs-docker-run/41918607#41918607

Comment (mrc0mmand) [reply ci/travis-centos.sh]: Also, here's a link to the issue: https://bugzilla.redhat.com/show_bug.cgi?id=1647343

File: ci/travis-centos.sh (hunk 2 above)
Comment (mrc0mmand) [reply ci/travis-centos.sh]: Here's the link for the issue: https://bugzilla.redhat.com/show_bug.cgi?id=1647343

File: ci/travis-centos.sh (hunk 1 above)
Comment (msekletar) [reply ci/travis-centos.sh]: Explanation on stackoverflow makes sense. Thanks for pointing it out!

File: ci/travis-centos.sh (hunk 2 above)
Comment (msekletar) [reply ci/travis-centos.sh]: Ok then, let's leave it as it is.

**Comment by mrc0mmand** [comment]
Current state of Travis for RHEL7/RHEL8:

### RHEL 7
- 4 failing tests, 3 of them should be fixed by https://github.com/lnykryn/systemd-rhel/pull/268
- the last one needs further debugging

### RHEL 8
- after merging https://github.com/lnykryn/systemd-rhel/pull/269 the Travis on RHEL 8 is fully functional
- known workarounds: `test/udev-test.pl` had to be disabled, as it relies on a functionality provided by newer `systemd-detect-virt` and it's pointless to run it in a virtualized docker container anyway

File: ci/travis-centos-rhel7.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            $DOCKER_EXEC ./autogen.sh
+            $DOCKER_EXEC ./configure --disable-timesyncd --disable-kdbus --disable-terminal \
+                                     --enable-gtk-doc --enable-compat-libs --disable-sysusers \
+                                     --disable-ldconfig --enable-lz4 --with-sysvinit-path=/etc/rc.d/init.d
+            $DOCKER_EXEC make
+            # Temporarily skip test_path_changed in src/test/test-path.c
+            # This particular test case timeouts on Ubuntu in a Docker image
+            # and needs further debugging (couldn't reproduce it on
+            # Fedora + Docker
+            $DOCKER_EXEC sed -i '/static void test_path_changed/,+18d' src/test/test-path.c
```
Comment (jsynacek) [inline ci/travis-centos-rhel7.sh]: Given that the sed command on the line following this one removes the actual usage of test_path_changed(), this line shouldn't be needed. Also, I'd say that the "+18d" is pretty fragile, something like "/^}\s*$/" would probably be better.

**Review by jsynacek — APPROVED** [review]
Otherwise, LGTM.

File: ci/travis-centos-rhel7.sh (hunk 4 above)
Comment (mrc0mmand) [reply ci/travis-centos-rhel7.sh]: > Given that the sed command on the line following this one removes the actual usage of test_path_changed(), this line shouldn't be needed. 

That was actually the original solution, however, due to the compilation options used, the definition had to be dropped as well:
```
src/test/test-path.c: At top level:
src/test/test-path.c:137:13: error: 'test_path_changed' defined but not used [-Werror=unused-function]
 static void test_path_changed(Manager *m) {
             ^
cc1: all warnings being treated as errors
```

> Also, I'd say that the "+18d" is pretty fragile, something like "/^}\s*$/" would probably be better.

That's definitely true, will change it.

**Pull request merged**